src/Features/Core/Portable/IntroduceVariable/IntroduceLocalForExpressionCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,102 @@
+        where TStatementSyntax : SyntaxNode
+        where TExpressionStatementSyntax : TStatementSyntax
+        where TLocalDeclarationStatementSyntax : TStatementSyntax
+    {
+        protected abstract bool IsValid(TExpressionStatementSyntax expressionStatement, TextSpan span);
+        protected abstract Task<TLocalDeclarationStatementSyntax> CreateLocalDeclarationAsync(Document document, TExpressionStatementSyntax expressionStatement, CancellationToken cancellationToken);
+
+        public sealed override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var expressionStatement = await GetExpressionStatementAsync(context).ConfigureAwait(false);
+            if (expressionStatement == null)
+            {
+                return;
+            }
+
+            var (document, _, cancellationToken) = context;
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var expression = syntaxFacts.GetExpressionOfExpressionStatement(expressionStatement);
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var type = semanticModel.GetTypeInfo(expression).Type;
+            if (type == null ||
+                type.SpecialType == SpecialType.System_Void)
+            {
+                return;
+            }
+
+            var singleLineExpression = syntaxFacts.ConvertToSingleLine(expression);
+            var nodeString = singleLineExpression.ToString();
+
+            context.RegisterRefactoring(new MyCodeAction(
+                string.Format(FeaturesResources.Introduce_local_for_0, nodeString),
+                c => IntroduceLocalAsync(document, expressionStatement, c)));
+        }
+
+        protected async Task<TExpressionStatementSyntax> GetExpressionStatementAsync(CodeRefactoringContext context)
+        {
+            var expressionStatement = await context.TryGetSelectedNodeAsync<TExpressionStatementSyntax>().ConfigureAwait(false);
+            if (expressionStatement == null)
+            {
+                // If an expression-statement wasn't selected, see if they're selecting
+                // an expression belonging to an expression-statement instead.
+                var expression = await context.TryGetSelectedNodeAsync<TExpressionSyntax>().ConfigureAwait(false);
+                expressionStatement = expression?.Parent as TExpressionStatementSyntax;
+            }
+
+            return expressionStatement != null && IsValid(expressionStatement, context.Span)
+                ? expressionStatement
+                : null;
+        }
+
+        private async Task<Document> IntroduceLocalAsync(
+            Document document, TExpressionStatementSyntax expressionStatement, CancellationToken cancellationToken)
+        {
+            var localDeclaration = await CreateLocalDeclarationAsync(document, expressionStatement, cancellationToken).ConfigureAwait(false);
```

Comment:
Is this something that we can add to the syntax generator or does it feel to specialized?

Reply:
I thought about it, but i wanted *really* fine-grained control over what was happening.  For example, if you have a semicolon token already, i wanted to make sure that was used exactly in the resultant local-decl.  Similarly, for VB, i did *not* want to generate a type, as it's very idiomaric to just say `dim x = new DateTime()` it just didn't feel like it was buying a lot to use the generator here.

Reply:
Can we just use [this](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Editing/SyntaxGenerator.cs,1452) API with null typeSymbol so it generates `var` or `Dim`? Anyways, it is minor issue as the language specific implementations are pretty trivial.

Reply:
we could... but i don't want 'var' if you aren't using 'prefer var' settings for C#.  and so on :)  basically, i could use it... but i'd still have to do work.  So i didn't bother :)   I can see about trying it out and seeing how much i save.

Reply:
Ok.  simplified it :)